tempesta_fw/sock.c

```diff
@@ -746,21 +746,22 @@ ss_tcp_process_skb(struct sock *sk, struct sk_buff *skb, int *processed)
 		 */
 		BUG_ON(conn == NULL);
 
-		if (SS_CONN_TYPE(sk) & Conn_Suspected) {
+		if (SS_CONN_TYPE(sk) & Conn_Stop) {
 			__kfree_skb(skb);
 			continue;
 		}
 
 		r = SS_CALL(connection_recv, conn, skb, offset);
 
+		if (SS_CONN_TYPE(sk) & Conn_Stop) {
+			TFW_DBG("[%d]: Stop processing: sk %p\n",
+				smp_processor_id(), sk);
+			break; /* connection stopped */
+		}
 		if (r < 0) {
 			TFW_DBG("[%d]: Processing error: sk %p r %d\n",
 			        smp_processor_id(), sk, r);
 			goto out; /* connection dropped */
-		} else if (r == SS_STOP) {
-			TFW_DBG("[%d]: Stop processing: sk %p\n",
-			        smp_processor_id(), sk);
-			break;
```

Comment:
I think that wasn't a good piece of code. The code takes place when we completely read a client request with `Connection: close`, but client sends us more data. We have to close the socket, i.e. send FIN and go through normal TCP close process with appropriate sequence numbers and ACKs. If we just `break` here, then we improperly update `tp->copied_seq`: we update it for the FIN, but not for other date after the request with `Connection: close`. Thus, connection closing can go with errors.

Please add a functional test which sends data like
```
GET / HTTP/1.1\r\n
Host: foo.com\r\n
\r\n
[many arbitrary data to fit couple skbs]
```

(One case with the tail data and one without the data). The result should be verified with tcpdump which must show that we have correct number of packets and there are proper FIN sequences. Ideally this should be more or less full connection closing test.

@intelfx and @ikoveshnikov please note the test scenarios: typically we have to develop many tests verifying correct operation of TCP.

Reply:
I am not sure I understand this. All skb that come here are counted in `tp->copied_seq`, so it's updated properly in all cases. That happens before they are passed to Tempesta for processing.

In this particular case the socket was not closed because we still had to send a response back the the client - when it came from a server. That's when a close would occur, according to the `Connection: close` header. That's was the whole purpose of `SS_STOP` compared to `SS_BLOCK`. `SS_BLOCK` would cause an immediate close, while `SS_STOP` would not.

All of that created a window for a misbehaving client that could continue sending more requests despite `Connection: close`. The window was between receiving that request and sending a response to it. In that case Tempesta would continue processing those newer requests. That was the deficiency of the old solution which I described in the comment to the issue.

Still, those new skb would be properly accounted for from TCP seq's point of view.

Closing would happen after

Reply:
>All skb that come here are counted in `tp->copied_seq`
If we `break` from loop in case of `TFW_STOP` (or ` Conn_Stop` flag detected) - then skbs which remain in `skb_list` (i.e. "unrolled" skbs) - will not be counted in `tp->copied_seq`.

Reply:
Skb in `skb_list` were part of a GRO skb. And that GRO skb had been fully counted before it was unrolled. I don't think there's need to count the data twice.

Reply:
The problem is with https://github.com/tempesta-tech/tempesta/blob/master/tempesta_fw/sock.c#L718 : when we account or not the TCP sequence numbers, so we need not to exit (break) before accounting all akb pieces. `skb_unroll()` doesn't know about TCP anything.

Reply:
Duh! Yes, indeed. Thanks for pointing that out.

It appears though, that other error code paths are also affected. That includes returning on error in `ss_skb_unroll()`, and breaking out of the loop on `(r < 0)`. `tp->copied_seq` would not be updated properly in these cases, and a potential FIN would not be counted as well.

In these cases the connection is closed immediately, and perhaps an improper `tp->copied_seq` doesn't really matter. But perhaps not, and then this should better be fixed as well in this patch.

Reply:
Yeah, good point: the other cases also treated in wrong way. But there is a difference: stopped connections still must be treated in graceful way (i.e. closed by FIN) while blocked connections should be reseted. We never cared about blocked connections, so this is why they are closed in current way. However, there is TODO comment https://github.com/tempesta-tech/tempesta/blob/master/tempesta_fw/sock.c#L495 and we actually should reset the bad connections. Created issue https://github.com/tempesta-tech/tempesta/issues/861 . @aleksostapenko keep in mind the issue during implementation of #860.